### DISCUSSION TITLE
Stress Testing Forward Node

### DISCUSSION THREAD

**Discussion by SVdetails** [Security-Onion-Solutions/securityonion#5695] [Unsupported Versions] [answered]
Has anyone ever stress tested a Forward (Sensor) node before?

We will be installing a Distributed Grid on a somewhat large network and I want to perform a little "stress test" on the sensor in order to see how much traffic it can handle before it starts to experience large amounts of packet loss.

I know SolarWinds has the "WAN Killer" application, and I am likely going to use this under its free trial, but I was just reaching out to the community to see if anyone has previously done this or if there are any recommendations.

The Forward Node is installed on bare-metal, if that helps.

Thanks in advance!

**Comment by mrocaruano** [comment, marked as answer]
We stress tested Security Onion 16.04's forward nodes using tcpreplay, and spare hardware. This basic hardware setup to stress test shouldn't change too much. (Everything changed in Security Onion 2.0 for monitoring and tuning tools, how to find suricata dropped packets data and where to configure cpu pinning, etc. so I’m not going to cover how to do that part, just how to generate the traffic to do a stress test in the first place) 

**You will need:** 
- A sending device that is equal to, or greater in specs as your forward node 
- PCAPs 

You will need loads of typical traffic (grab them from a forward node if you're keeping full pcaps.  If you're not keeping them, consider storing some at least long enough to get a good snapshot of your typical traffic). Then grab some malware sample pcaps, and other non-typical traffic you might encounter, you'll want to test a wide variety of these as well. Here are some useful links for getting pcaps of both current and past threats:   

https://www.malware-traffic-analysis.net/ 
https://www.netresec.com/?page=PcapFiles 

Make sure you have several gigabytes of pcaps and a WIDE variety of traffic. You’ll need to separate your pcap files into directories with a total size less than the total amount of free RAM in your sending device -- you're going to need to preload all the pcaps into memory to ensure they’re sending at the bandwidth you're testing. The more variety you have, the better your tests will be. 

**Hardware Setup** 

You need a sending device with as much CPU and RAM as you can get (MINIMUM equal to the forward node, but preferably more), and a network interface equal to, or greater than the forward node's sniffer interface. This will be very resource intensive on the sending device so make sure it can handle the stress too! 

You'll need to run a point to point network cable between the sending device and your forward node. Type of cable is dependent on the NIC you're testing (For example, we did this with 10Gb fiber NICs on both ends). 

Your sending device needs Linux installed - any flavor should do. Copy all of your pcaps to a single directory on your sending device (or multiple directories if you have more pcaps than RAM). Then, from your pcap directory, you can use this tcpdump command, edited as needed, to simulate traffic at various speeds: 

`$ sudo tcpreplay --preload-pcap --intf1=<interface> --mbps=<speed> --loop=<number of times to repeat> *.pcap `  

We used 10-20gb worth of pcaps  per test, on loops of 100/1000 or more so they could run for hours/days when needed. We would just stop the feeds with ctrl-C when/if needed. While tcpreplay is running you can monitor your forward node and make whatever tuning adjustments you need. Repeat, cycling through your directories of pcaps, until you’re happy with your tuning!

**Reply by SVdetails** [reply]
mrocaruano,

Thanks so much for the prompt and detailed reply. 

I was actually reading up on the different ways to perform the stress test and some of them were in line with your methods (though not as thorough). 

The most difficult part now is to find hardware with higher specs than our current sensor hardware. We're essentially using our best hardware already!!

Thanks again brother. 🍻
